File: typescript/top-interview-150/189-rotate-array/main.ts
function rotate(nums: number[], k: number): void {
  const n = nums.length;
  k = ((k % n) + n) % n; // normalize to [0, n)

  const countGCD = (a: number, b: number): number =>
    b === 0 ? a : countGCD(b, a % b);

  const cycles = countGCD(n, k);

  for (let start = 0; start < cycles; start++) {
    let current = start;
    let prev = nums[start];

    while (true) {
      const next = (current + k) % n;
      const temp = nums[next];
      nums[next] = prev;
      prev = temp;
      current = next;

      if (current === start) break;
    }
  }
}


const expected: number[] = []
const nums: number[] = []
const k: number = null as unknown as number;
rotate(nums, k);
console.log(nums);

let correct = true;
for (let i = 0; i < nums.length; i++) {
	if (nums[i] !== expected[i]) {
		correct = false;
		break;
	}
}
console.log(correct);

File: python/top-interview-150/189-rotate-array/main.py
from typing import List


class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        rotated = []
        offset = (len(nums) - k) % len(nums)
        for x in nums[offset : len(nums)]:
            rotated.append(x)
        for x in nums[0:offset]:
            rotated.append(x)

        for i, x in enumerate(rotated):
            nums[i] = x


nums = [1, 2, 3, 4, 5, 6, 7]
k = 3
expected = [5, 6, 7, 1, 2, 3, 4]

nums = [-1, -100, 3, 99]
k = 2
expected = [3, 99, -1, -100]

Solution().rotate(nums=nums, k=k)
print(f"{nums}")
print(nums == expected)
